### PR TITLE
Assert every VS Code extension E2E spec has a CI matrix row

### PR DESCRIPTION
**Pull request by adamint**
## Description

Adds a unit test asserting that the VS Code extension E2E specs and the `extension-e2e-tests.yml` matrix match exactly in both directions.

A new `src/test-e2e/*.e2e.test.ts` file without a matrix row is otherwise never scheduled, while a stale row can run zero tests. Neither failure is visible from inside the E2E suite, so this guard runs with the normal extension unit tests.

The workflow is parsed with the extension's existing `js-yaml` dependency rather than a handwritten YAML parser. The guard then checks:

- spec discovery is recursive, matching Mocha's E2E configuration;
- every matrix row has a non-empty `spec`;
- duplicate Linux/Windows rows for one spec are allowed;
- nested or unrelated `spec` properties do not count as `matrix.spec`;
- disabled rows exactly match an explicit allowlist and use a microsoft/aspire issue URL.

Split out of #19133, where the resource-debugger harness first needed this coverage.

## Validation

```console
corepack yarn --silent compile-tests
corepack yarn --silent lint
./node_modules/.bin/mocha out/test/e2eShardMatrix.test.js --ui tdd

  9 passing
```

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
- Did you add public API?
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [x] No

### CODE REVIEW

File: extension/src/test/e2eShardMatrix.test.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    test('points every CI matrix row at a spec that exists', () => {
+        const workflow = fs.readFileSync(workflowPath, 'utf8');
+        const matrixSpecs = [...new Set([...workflow.matchAll(/^\s*spec:\s*(\S+)\s*$/gm)].map(match => match[1]))].sort();
+
+        assert.ok(matrixSpecs.length > 0, 'Expected spec entries in the E2E workflow matrix.');
+        // The runner resolves the spec with a glob, and a glob that matches nothing yields a shard
+        // that runs zero tests, so a renamed or deleted spec has to be caught here as well.
+        assert.deepStrictEqual(
+            matrixSpecs.filter(spec => !fs.existsSync(path.join(extensionRoot, spec.replace('out/test-e2e/test-e2e/', 'src/test-e2e/').replace(/\.js$/, '.ts')))),
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/test/e2eShardMatrix.test.ts]: Fixed in 562fdc3. The two directional checks are now one `deepStrictEqual` between the deduplicated matrix values and the compiled paths derived from the `.e2e.test.ts` files, so a helper is no longer a valid value. Injecting `out/test-e2e/test-e2e/helpers/fixtures.js` passed both old tests and fails the new one.

[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/test/e2eShardMatrix.test.ts]: Added the missing proof in `d1b38aaa6d`: `canonicalSpecPaths`/`matrixSpecPaths` now take spec names and workflow text as parameters, so synthetic matrices can be asserted against. The old existence check returned zero offenders for a `helpers/fixtures.js` row; set equality rejects it, and injecting that row into the real workflow fails `runs exactly the set of E2E specs in the CI matrix`.

File: extension/src/test/e2eShardMatrix.test.ts [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    /**
+     * Matrix rows carry the compiled spec path:
+     *       - name: Linux
+     *         shardName: edge-cases
+     *         spec: out/test-e2e/test-e2e/edgeCases.e2e.test.js
+     * A spec legitimately appears on more than one row (one per platform), so the values are
+     * deduplicated before being compared with the canonical set.
+     */
+    function matrixSpecPaths(workflow: string): string[] {
+        return [...new Set([...workflow.matchAll(/^\s*spec:\s*(\S+)\s*$/gm)].map(match => match[1]))].sort();
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/test/e2eShardMatrix.test.ts]: Valid, and fixed at `542686611f75859668b2c079a10aeefcb1d924ab`.

`matrixSpecPaths` now runs the captured value through the same `parseWorkflowScalar` helper the other matrix fields use, so the quotes are stripped before the comparison (`extension/src/test/e2eShardMatrix.test.ts:98`):

```ts
return [...new Set([...matrixIncludeBlock(workflow).matchAll(/^\s*spec:\s*(\S+)\s*$/gm)].map(match => parseWorkflowScalar(match[1])))].sort();
```

`setMatrixRowField` routes through the same helper, so a quoted `spec` parses identically on the row path and the set-comparison path.

The exact row from the finding is pinned by `accepts a quoted spec scalar`:

```ts
test('accepts a quoted spec scalar', () => {
    const workflow = workflowWithSpecs("'out/test-e2e/test-e2e/edgeCases.e2e.test.js'");

    assertMatrixMatchesSpecs(workflow, ['edgeCases.e2e.test.ts']);
});
```

Mutation-proved rather than assumed. Reverting only the helper call in `matrixSpecPaths` to `.map(match => match[1])`, rebuilding, and re-running under `vscode-test`:

| state | result |
|---|---|
| at head | **12 passing, 0 failing** |
| `parseWorkflowScalar` removed from `matrixSpecPaths` | **11 passing, 1 failing** — `accepts a quoted spec scalar`, `AssertionError: The spec values in extension-e2e-tests.yml must be exactly the compiled paths of the .e2e.test.ts files under src/test-e2e.` |
| restored | **12 passing, 0 failing** |

The test fails only in the reverted state, so it pins the behaviour rather than passing vacuously.

File: extension/src/test/e2eShardMatrix.test.ts [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+
+        appendCurrentRow();
+
+        return rows;
+
+        function appendCurrentRow(): void {
+            if (currentRow?.spec !== undefined || currentRow?.disabledIssue !== undefined) {
+                rows.push(currentRow);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/test/e2eShardMatrix.test.ts]: Real, and it was still real at `195b10a2` after the `assert.ok(row.spec, ...)` change. Closed in `7c69c8d48cf8d49399ab1af4a13eea47ef932d6c`.

**What was left open.** The assertion only sees rows `matrixRows` produces, and the row pattern is `/^(\s*)-\s+([A-Za-z0-9_-]+):\s*(.*?)\s*$/`. Two row shapes GitHub Actions accepts never match it. I ran both against the guard as it stood at `195b10a2`:

```
PROBE bare-dash rows = [{"name":"Windows","shardName":"edge-cases","spec":"out/test-e2e/test-e2e/edgeCases.e2e.test.js"}]
PROBE bare-dash => ACCEPTED (bypass)
PROBE flow rows = [{"name":"Linux","shardName":"edge-cases","spec":"out/test-e2e/test-e2e/edgeCases.e2e.test.js"}]
PROBE flow => ACCEPTED (bypass)
PROBE empty-spec => REJECTED: E2E matrix rows must include spec.
PROBE commented => REJECTED: E2E matrix rows must include spec.
```

Input for the first probe was a Linux row with a spec followed by

```yaml
          -
            name: Windows
            shardName: edge-cases
            runner: windows-latest
```

The flow-style row (`- { name: Windows, shardName: edge-cases }`) is dropped outright. The bare-dash row is worse than a drop: the parsed output above is a single row, because the Windows keys land on the still-open Linux row and overwrite `name` and `shardName` while inheriting the Linux `spec`. The spec-less row satisfies the assertion using another row's value.

The fix treats any sequence entry at the row indent that the row pattern does not match as a failure rather than skipping it. Nested sequences under a row property sit deeper than the row indent, so they are unaffected; there is an accept-direction test for that.

**Counts.** Suite is `extension/src/test/e2eShardMatrix.test.ts`, run as `tsc -p . --outDir out` then `mocha --ui tdd out/test/e2eShardMatrix.test.js`.

| state | passing | failing |
|---|---|---|
| `195b10a2` (before this commit) | 14 | 0 |
| `7c69c8d4` (fix + 4 tests) | 18 | 0 |
| M1: drop the fail-closed branch | 16 | 2 — bare-dash, flow-style |
| M2: revert `195b10a2`'s `assert.ok(row.spec)` and the unconditional `appendCurrentRow` | 16 | 2 — row-without-spec, empty-spec |
| M3: make `assertMatrixMatchesSpecs` a no-op | 9 | 9 |

M3 is the discriminating-power check, since a fail-open guard passes a plain revert trivially. Neutering the comparison fails all nine reject cases and passes all nine accept cases, including `runs exactly the set of E2E specs in the CI matrix` against the real workflow file. The two `disabled matrix row` reject cases are in the passing nine because they go through `assertDisabledRowsAreTracked`, not the neutered function.

**One correction to the downstream chain.** A spec-less row does not reach the fallback glob in CI. The job sets `defaults.run.shell: pwsh` and the `Compile extension E2E tests` step runs `if (-not (Test-Path '${{ matrix.spec }}')) { throw ... }`. With an absent or empty `matrix.spec` that renders as `Test-Path ''`, which returns false rather than erroring, so the step's own `throw` fires and the shard fails. Verified locally under `$ErrorActionPreference = 'stop'`:

```
RESULT: terminating error => RuntimeException: compiled spec not found
```

So the `run-e2e.js` fallback to `out/test-e2e/**/*.e2e.test.js` is the counterfactual, not the observed CI behaviour. That does not weaken the finding: this guard is the cheap pre-CI signal, and the point stands that it could not fail on the input it exists to reject.

File: extension/src/test/e2eShardMatrix.test.ts [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+                if (rowIndent !== includeRowIndent) {
+                    continue;
+                }
+
+                appendCurrentRow();
+                currentRow = {};
+                currentRowIndent = rowIndent;
+                setMatrixRowField(currentRow, rowStart[2], rowStart[3]);
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/test/e2eShardMatrix.test.ts]: At current head `7c69c8d`, this now fails closed instead of silently ignoring that YAML shape. `extension/src/test/e2eShardMatrix.test.ts` checks sequence entries at the include-row indent and calls `assert.fail` for unsupported row shapes (the bare-dash and flow-style cases), with regression tests `rejects a matrix row whose keys start on the line after the dash` and `rejects a flow-style matrix row`. So a standalone-dash row cannot pass the guard or inherit the previous row spec.

**Review by ellahathaway — APPROVED** [review]
